File: backend/src/main/java/com/iotplatform/service/RuleService.java
package com.iotplatform.service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.iotplatform.controller.RuleController;
import com.iotplatform.dto.TelemetryDataRequest;
import com.iotplatform.dto.RulesGenerationResponse;
import com.iotplatform.model.Device;
import com.iotplatform.model.Rule;
import com.iotplatform.model.RuleAction;
import com.iotplatform.model.RuleCondition;
import com.iotplatform.repository.DeviceRepository;
import com.iotplatform.repository.RuleRepository;

@Service
public class RuleService {

    @Autowired
    private RuleRepository ruleRepository;

    @Autowired
    private DeviceRepository deviceRepository;

    @Autowired
    private NotificationService notificationService;

    public List<Rule> getAllRules(String organizationId) {
        return ruleRepository.findByOrganizationId(organizationId);
    }

    public Optional<Rule> getRule(String id, String organizationId) {
        return ruleRepository.findByIdAndOrganizationId(id, organizationId);
    }

    public Rule createRule(Rule rule, String organizationId) {
        rule.setId(UUID.randomUUID().toString());
        rule.setOrganizationId(organizationId);
        
        // Log rule creation with device ID
        System.out.println("RuleService: Creating rule '" + rule.getName() + "' for device: " + rule.getDeviceId());
        
        Rule savedRule = ruleRepository.save(rule);
        
        // Log successful save
        System.out.println("RuleService: Successfully saved rule '" + savedRule.getId() + "' for device: " + savedRule.getDeviceId());
        
        return savedRule;
    }

    public List<Rule> createBulkRules(List<java.util.Map<String, Object>> rulesData, String organizationId) {
        List<Rule> createdRules = new java.util.ArrayList<>();
        
        for (java.util.Map<String, Object> ruleData : rulesData) {
            try {
                Rule rule = new Rule();
                rule.setId(UUID.randomUUID().toString());
                rule.setName((String) ruleData.get("name"));
                rule.setDescription((String) ruleData.get("description"));
                
                // Set new fields if available
                if (ruleData.containsKey("metric")) {
                    rule.setMetric((String) ruleData.get("metric"));
                }
                if (ruleData.containsKey("metricValue")) {
                    rule.setMetricValue((String) ruleData.get("metricValue"));
                }
                if (ruleData.containsKey("threshold")) {
                    rule.setThreshold((String) ruleData.get("threshold"));
                }
                if (ruleData.containsKey("consequence")) {
                    rule.setConsequence((String) ruleData.get("consequence"));
                }
                if (ruleData.containsKey("deviceId")) {
                    rule.setDeviceId((String) ruleData.get("deviceId"));
                }
                
                rule.setActive(true);
                rule.setOrganizationId(organizationId);
                
                // Create simple condition and action based on the data
                if (ruleData.containsKey("condition") && ruleData.containsKey("action")) {
                    List<RuleCondition> conditions = new java.util.ArrayList<>();
                    List<RuleAction> actions = new java.util.ArrayList<>();
                    
                    // Create condition
                    RuleCondition condition = new RuleCondition();
                    condition.setId(UUID.randomUUID().toString());
                    condition.setType(RuleCondition.ConditionType.TELEMETRY_THRESHOLD);
                    condition.setMetric("condition");
                    condition.setOperator(RuleCondition.Operator.EQUALS);
                    condition.setValue((String) ruleData.get("condition"));
                    condition.setLogicOperator(RuleCondition.LogicOperator.AND);
                    condition.setRule(rule);
                    
                    // Set device ID if provided
                    if (ruleData.containsKey("deviceId")) {
                        condition.setDeviceId((String) ruleData.get("deviceId"));
                    }
                    
                    conditions.add(condition);
                    
                    // Create action
                    RuleAction action = new RuleAction();
                    action.setId(UUID.randomUUID().toString());
                    action.setType(RuleAction.ActionType.NOTIFICATION);
                    
                    // Set action config
                    java.util.Map<String, String> config = new java.util.HashMap<>();
                    config.put("message", (String) ruleData.get("action"));
                    if (ruleData.containsKey("priority")) {
                        config.put("priority", (String) ruleData.get("priority"));
                    }
                    if (ruleData.containsKey("category")) {
                        config.put("category", (String) ruleData.get("category"));
                    }
                    action.setConfig(config);
                    action.setRule(rule);
                    
                    actions.add(action);
                    
                    rule.setConditions(conditions);
                    rule.setActions(actions);
                }
                
                Rule savedRule = ruleRepository.save(rule);
                createdRules.add(savedRule);
                
            } catch (Exception e) {
                System.err.println("Failed to create rule: " + e.getMessage());
                e.printStackTrace();
                // Continue with other rules
            }
        }
        
        return createdRules;
    }

    public Rule updateRule(String id, Rule ruleDetails, String organizationId) {
        Rule rule = ruleRepository.findByIdAndOrganizationId(id, organizationId)
                .orElseThrow(() -> new RuntimeException("Rule not found"));

        rule.setName(ruleDetails.getName());
        rule.setDescription(ruleDetails.getDescription());
        rule.setMetric(ruleDetails.getMetric());
        rule.setMetricValue(ruleDetails.getMetricValue());
        rule.setThreshold(ruleDetails.getThreshold());
        rule.setConsequence(ruleDetails.getConsequence());
        rule.setDeviceId(ruleDetails.getDeviceId());
        rule.setActive(ruleDetails.isActive());
        rule.setConditions(ruleDetails.getConditions());
        rule.setActions(ruleDetails.getActions());

        return ruleRepository.save(rule);
    }

    public void deleteRule(String id, String organizationId) {
        Rule rule = ruleRepository.findByIdAndOrganizationId(id, organizationId)
                .orElseThrow(() -> new RuntimeException("Rule not found"));
        ruleRepository.delete(rule);
    }

    public List<Rule> getActiveRules(String organizationId) {
        return ruleRepository.findByOrganizationIdAndActive(organizationId, true);
    }

    public List<Rule> getRulesByDevice(String deviceId, String organizationId) {
        return ruleRepository.findByDeviceIdAndOrganizationId(deviceId, organizationId);
    }

    public RuleController.RuleStats getRuleStats(String organizationId) {
        long totalRules = ruleRepository.findByOrganizationId(organizationId).size();
        long activeRules = ruleRepository.findByOrganizationIdAndActive(organizationId, true).size();
        
        // For now, return placeholder values for triggered counts
        // In a real implementation, you would query actual trigger history
        long triggeredToday = 0;
        long triggeredThisWeek = 0;
        
        return new RuleController.RuleStats(totalRules, activeRules, triggeredToday, triggeredThisWeek);
    }

    public void evaluateRules(String deviceId, TelemetryDataRequest telemetryData, String organizationId) {
        // Get device
        Device device = deviceRepository.findByIdAndOrganizationId(deviceId, organizationId)
                .orElse(null);
        
        if (device == null) return;

        // Get all rules for this specific device
        List<Rule> deviceRules = ruleRepository.findByDeviceIdAndOrganizationId(deviceId, organizationId);

        for (Rule rule : deviceRules) {
            // Only evaluate active rules
            if (!rule.isActive()) continue;
            if (evaluateRuleConditions(rule, telemetryData)) {
                executeRuleActions(rule, device, telemetryData);
                
                // Update last triggered timestamp
                rule.setLastTriggered(LocalDateTime.now());
                ruleRepository.save(rule);
            }
        }
    }

    private boolean evaluateRuleConditions(Rule rule, TelemetryDataRequest telemetryData) {
        // Enhanced rule evaluation using new metric and threshold fields
        if (rule.getMetric() != null && rule.getThreshold() != null) {
            // Use the new metric and threshold fields from the rule
            String metric = rule.getMetric();
            if (telemetryData.getMetrics().containsKey(metric)) {
                double actualValue = telemetryData.getMetrics().get(metric);
                
                // Parse threshold value (e.g., "2.5 mm/s" -> 2.5)
                String thresholdStr = rule.getThreshold();
                try {
                    // Extract numeric value from threshold string
                    String numericPart = thresholdStr.replaceAll("[^0-9.]", "");
                    if (!numericPart.isEmpty()) {
                        double thresholdValue = Double.parseDouble(numericPart);
                        
                        // Determine operator from threshold string
                        if (thresholdStr.contains(">")) {
                            return actualValue > thresholdValue;
                        } else if (thresholdStr.contains("<")) {
                            return actualValue < thresholdValue;
                        } else if (thresholdStr.contains("=")) {
                            return actualValue == thresholdValue;
                        } else if (thresholdStr.contains(">=")) {
                            return actualValue >= thresholdValue;
                        } else if (thresholdStr.contains("<=")) {
                            return actualValue <= thresholdValue;
                        }
                    }
                } catch (NumberFormatException e) {
                    System.err.println("Failed to parse threshold value: " + thresholdStr);
                }
            }
        }
        
        // Fallback to original condition evaluation
        for (RuleCondition condition : rule.getConditions()) {
            if (condition.getType() == RuleCondition.ConditionType.TELEMETRY_THRESHOLD) {
                String metric = condition.getMetric();
                if (telemetryData.getMetrics().containsKey(metric)) {
                    double actualValue = telemetryData.getMetrics().get(metric);
                    double expectedValue = Double.parseDouble(condition.getValue());
                    
                    return switch (condition.getOperator()) {
                        case GREATER_THAN -> actualValue > expectedValue;
                        case LESS_THAN -> actualValue < expectedValue;
                        case EQUALS -> actualValue == expectedValue;
                        case GREATER_THAN_OR_EQUAL -> actualValue >= expectedValue;
                        case LESS_THAN_OR_EQUAL -> actualValue <= expectedValue;
                    };
                }
            }
        }
        return false;
    }

    private void executeRuleActions(Rule rule, Device device, TelemetryDataRequest telemetryData) {
        // Simplified action execution - in production, implement all action types
        System.out.println("Executing actions for rule: " + rule.getName());
        System.out.println("Device: " + device.getName());
        System.out.println("Telemetry: " + telemetryData.getMetrics());
        
        // Create notification
        try {
            notificationService.createRuleTriggeredNotification(rule, device, telemetryData);
        } catch (Exception e) {
            System.err.println("Failed to create notification: " + e.getMessage());
        }
    }
    
    public List<Rule> getRulesByDeviceId(String deviceId) {
        return ruleRepository.findByDeviceId(deviceId);
    }
    
    public void createRulesFromPDF(List<RulesGenerationResponse.Rule> rules, String deviceId, String organizationId) {
        System.out.println("Creating rules from PDF for device: " + deviceId);
        for (RulesGenerationResponse.Rule ruleData : rules) {
            Rule rule = new Rule();
            rule.setId(UUID.randomUUID().toString());
            rule.setName(ruleData.getName());
            rule.setDescription(ruleData.getDescription());
            rule.setMetric(ruleData.getMetric());
            rule.setMetricValue(ruleData.getMetricValue());
            rule.setThreshold(ruleData.getThreshold());
            rule.setConsequence(ruleData.getConsequence());
            rule.setActive(true);
            rule.setDeviceId(deviceId);
            rule.setOrganizationId(organizationId);
            rule.setCreatedAt(LocalDateTime.now());
            rule.setUpdatedAt(LocalDateTime.now());
            
            ruleRepository.save(rule);
        }
        System.out.println("Created " + rules.size() + " rules from PDF for device: " + deviceId);
    }
}

File: server/services/ruleService.ts
import { supabase } from '../config/database';
import { v4 as uuidv4 } from 'uuid';
import { notificationService } from './notificationService';
import { deviceService } from './deviceService';

export interface Rule {
  id: string;
  name: string;
  description: string;
  active: boolean;
  conditions: RuleCondition[];
  actions: RuleAction[];
  organization_id: string;
  created_at: string;
  updated_at: string;
  last_triggered?: string;
}

export interface RuleCondition {
  id: string;
  type: 'device_status' | 'telemetry_threshold' | 'time_based';
  device_id?: string;
  metric?: string;
  operator: '>' | '<' | '=' | '>=' | '<=';
  value: string | number;
  logic_operator?: 'AND' | 'OR';
}

export interface RuleAction {
  id: string;
  type: 'notification' | 'device_control' | 'webhook' | 'log';
  config: Record<string, any>;
}

export class RuleService {
  async createRule(ruleData: Partial<Rule>, organizationId: string): Promise<Rule> {
    const rule = {
      id: uuidv4(),
      ...ruleData,
      organization_id: organizationId,
      created_at: new Date().toISOString(),
      updated_at: new Date().toISOString()
    };

    const { data, error } = await supabase
      .from('rules')
      .insert([rule])
      .select()
      .single();

    if (error) throw error;
    return data;
  }

  async getRules(organizationId: string): Promise<Rule[]> {
    const { data, error } = await supabase
      .from('rules')
      .select('*')
      .eq('organization_id', organizationId)
      .order('created_at', { ascending: false });

    if (error) throw error;
    return data || [];
  }

  async getRule(ruleId: string, organizationId: string): Promise<Rule | null> {
    const { data, error } = await supabase
      .from('rules')
      .select('*')
      .eq('id', ruleId)
      .eq('organization_id', organizationId)
      .single();

    if (error) {
      if (error.code === 'PGRST116') return null;
      throw error;
    }

    return data;
  }

  async updateRule(ruleId: string, updates: Partial<Rule>, organizationId: string): Promise<Rule> {
    const { data, error } = await supabase
      .from('rules')
      .update({
        ...updates,
        updated_at: new Date().toISOString()
      })
      .eq('id', ruleId)
      .eq('organization_id', organizationId)
      .select()
      .single();

    if (error) throw error;
    return data;
  }

  async deleteRule(ruleId: string, organizationId: string): Promise<void> {
    const { error } = await supabase
      .from('rules')
      .delete()
      .eq('id', ruleId)
      .eq('organization_id', organizationId);

    if (error) throw error;
  }

  async evaluateRules(deviceId: string, telemetryData: any): Promise<void> {
    // Get device info
    const device = await deviceService.getDevice(deviceId, telemetryData.organization_id);
    if (!device) return;

    // Get active rules for this organization
    const rules = await this.getRules(device.organization_id);
    const activeRules = rules.filter(rule => rule.active);

    for (const rule of activeRules) {
      const shouldTrigger = await this.evaluateRuleConditions(rule, device, telemetryData);
      
      if (shouldTrigger) {
        await this.executeRuleActions(rule, device, telemetryData);
        
        // Update last triggered timestamp
        await this.updateRule(rule.id, {
          last_triggered: new Date().toISOString()
        }, device.organization_id);
      }
    }
  }

  private async evaluateRuleConditions(rule: Rule, device: any, telemetryData: any): Promise<boolean> {
    const results: boolean[] = [];

    for (const condition of rule.conditions) {
      let result = false;

      switch (condition.type) {
        case 'device_status':
          result = this.evaluateStatusCondition(condition, device);
          break;
        case 'telemetry_threshold':
          result = this.evaluateTelemetryCondition(condition, telemetryData);
          break;
        case 'time_based':
          result = this.evaluateTimeCondition(condition);
          break;
      }

      results.push(result);
    }

    // Evaluate logic operators (simplified - assumes all AND for now)
    return results.every(result => result);
  }

  private evaluateStatusCondition(condition: RuleCondition, device: any): boolean {
    const deviceStatus = device.status;
    const expectedStatus = condition.value;

    switch (condition.operator) {
      case '=':
        return deviceStatus === expectedStatus;
      case '!=':
        return deviceStatus !== expectedStatus;
      default:
        return false;
    }
  }

  private evaluateTelemetryCondition(condition: RuleCondition, telemetryData: any): boolean {
    if (!condition.metric || !telemetryData.metrics) return false;

    const actualValue = telemetryData.metrics[condition.metric];
    const expectedValue = parseFloat(condition.value.toString());

    if (actualValue === undefined || isNaN(expectedValue)) return false;

    switch (condition.operator) {
      case '>':
        return actualValue > expectedValue;
      case '<':
        return actualValue < expectedValue;
      case '=':
        return actualValue === expectedValue;
      case '>=':
        return actualValue >= expectedValue;
      case '<=':
        return actualValue <= expectedValue;
      default:
        return false;
    }
  }

  private evaluateTimeCondition(condition: RuleCondition): boolean {
    // Simplified time-based evaluation
    const now = new Date();
    const hour = now.getHours();
    
    // Example: trigger between 9 AM and 5 PM
    return hour >= 9 && hour <= 17;
  }

  private async executeRuleActions(rule: Rule, device: any, telemetryData: any): Promise<void> {
    for (const action of rule.actions) {
      try {
        switch (action.type) {
          case 'notification':
            await this.executeNotificationAction(action, rule, device, telemetryData);
            break;
          case 'device_control':
            await this.executeDeviceControlAction(action, device);
            break;
          case 'webhook':
            await this.executeWebhookAction(action, rule, device, telemetryData);
            break;
          case 'log':
            await this.executeLogAction(action, rule, device, telemetryData);
            break;
        }
      } catch (error) {
        console.error(`Failed to execute action ${action.type} for rule ${rule.id}:`, error);
      }
    }
  }

  private async executeNotificationAction(action: RuleAction, rule: Rule, device: any, telemetryData: any): Promise<void> {
    const notification = {
      title: `Rule Triggered: ${rule.name}`,
      message: `Device ${device.name} has triggered rule "${rule.name}". Current status: ${device.status}`,
      type: 'warning' as const,
      device_id: device.id,
      rule_id: rule.id,
      organization_id: device.organization_id
    };

    await notificationService.createNotification(notification);

    // Send to configured channels
    if (action.config.channels) {
      for (const channel of action.config.channels) {
        await notificationService.sendNotification(notification, channel);
      }
    }
  }

  private async executeDeviceControlAction(action: RuleAction, device: any): Promise<void> {
    // Implement device control logic
    console.log(`Executing device control action for device ${device.id}:`, action.config);
  }

  private async executeWebhookAction(action: RuleAction, rule: Rule, device: any, telemetryData: any): Promise<void> {
    // Implement webhook execution
    const payload = {
      rule: rule.name,
      device: device.name,
      telemetry: telemetryData,
      timestamp: new Date().toISOString()
    };

    console.log(`Executing webhook action:`, payload);
  }

  private async executeLogAction(action: RuleAction, rule: Rule, device: any, telemetryData: any): Promise<void> {
    // Log the rule execution
    console.log(`Rule ${rule.name} triggered for device ${device.name}:`, {
      rule: rule.id,
      device: device.id,
      telemetry: telemetryData,
      timestamp: new Date().toISOString()
    });
  }
}

export const ruleService = new RuleService();